bitbots_hcm/bitbots_hcm/humanoid_control_module.py

```diff
@@ -1,203 +1,195 @@
+        """ Updates the power state. """
         self.blackboard.is_power_on = msg.data
 
-    def get_state(self):
+    def diag_cb(self, msg: DiagnosticArray):
+        """ Updates the diagnostic state. """
+        status: DiagnosticStatus
+        for status in msg.status:
+            if "//Servos/" in status.name:
+                if status.level == DiagnosticStatus.ERROR and "Overload" in status.message:
+                    self.blackboard.servo_overload = True
+            elif "//Servos" in status.name:
+                self.blackboard.servo_diag_error = status.level in (DiagnosticStatus.ERROR, DiagnosticStatus.STALE)
+            elif "//IMU" in status.name:
+                self.blackboard.imu_diag_error = status.level in (DiagnosticStatus.ERROR, DiagnosticStatus.STALE)
+            elif "//Pressure" in status.name:
+                self.blackboard.pressure_diag_error = status.level in (DiagnosticStatus.ERROR, DiagnosticStatus.STALE)
+
+    def get_state(self) -> RobotControlState:
+        """ Returns the current state of the HCM. """
         return self.blackboard.current_state
 
-    def loop(self):
-        """ Keeps updating the DSD and publish its current state.
-            All the forwarding of joint goals is directly done in the callbacks to reduce latency. """
-        if self.blackboard.shut_down_request and not self.simulation_active:
-            self.on_shutdown_hook()
-        else:
-            loop_start_time = self.node.get_clock().now()
-            #can happen in simulation due to bad implementation in rclpy
-            if (self.last_loop_start_time != loop_start_time):
-                self.last_loop_start_time = loop_start_time
-                if self.hcm_deactivated:
-                    self.blackboard.current_state = RobotControlState.CONTROLLABLE
-                else:
-                    self.blackboard.current_time = self.node.get_clock().now()
-                    try:
-                        self.dsd.update()
-                    except IndexError:
-                        # this error will happen during shutdown procedure, just ignore it
-                        pass
-
-    def on_shutdown_hook(self):
-        if not self.blackboard:
-            return
-        # we got external shutdown, tell it to the DSD, it will handle it
-        self.blackboard.shut_down_request = True
-        self.node.get_logger().warn("You're stopping the HCM. The robot will sit down and power off its motors.")
-        speak("Stopping HCM", self.blackboard.speak_publisher, priority=50)
-        # now wait for it finishing the shutdown procedure
-        while not self.blackboard.current_state == RobotControlState.HCM_OFF:
-            # we still have to update everything
-            self.blackboard.current_time = self.node.get_clock().now()
-            self.dsd.update()
-            self.hcm_state_publisher.publish(self.blackboard.current_state)
-            self.node.get_clock().sleep_for(Duration(seconds=0.01))
```

Comment:
Is this handled somewhere else or removed entirely?

Reply:
It is not needed anymore, as we are unable to perform a sit down motion.